managers/backup-manager/BackupManager.js

```diff
@@ -0,0 +1,421 @@
+        tenant_id: opts.context ? this.getTenantGuid(opts.context) : args.space_guid,
+        description: `${backup.trigger} backup triggerred by ${data.username} at ${data.started_at}`
+      })
+      .value();
+
+    function createSecret() {
+      return utils
+        .randomBytes(12)
+        .then(buffer => buffer.toString('base64'));
+    }
+
+    function normalizeVm(vm) {
+      let vmParams = _.pick(vm, 'cid', 'agent_id', 'job', 'index');
+      return _.set(vmParams, 'iaas_vm_metadata.vm_id', config.backup.provider.name === CONST.IAAS.AZURE ? vmParams.agent_id : vmParams.cid);
+    }
+
+    let metaUpdated = false,
+      backupStarted = false,
+      registeredStatusPoller = false;
+
+    let deploymentName;
+    return Promise.try(() => {
+        return BackupManager
+          .findNetworkSegmentIndex(opts.instance_guid)
+          .then(networkIndex => BackupManager.getDeploymentName(opts.instance_guid, networkIndex))
+          .then(res => {
+            deploymentName = res;
+            logger.info('Obtained the deployment name for instance :', deploymentName);
+          });
+      })
+      .then(() => Promise.all([
+        createSecret(),
+        this.getDeploymentIps(deploymentName),
+        this.director.getDeploymentVms(deploymentName).map(normalizeVm)
+      ]))
+      .spread((secret, ips, vms) => {
+        // set data and backup secret
+        logger.info(`Starting backup on - ${deploymentName}. Agent Ips for deployment - `, ips);
+        data.secret = backup.secret = secret;
+        return this.agent
+          .getHost(ips, 'backup')
+          .tap(agent_ip => {
+            data.agent_ip = result.agent_ip = agent_ip;
+            instanceInfo = _.chain(data)
+              .pick('tenant_id', 'backup_guid', 'instance_guid', 'agent_ip', 'service_id', 'plan_id')
+              .set('deployment', deploymentName)
+              .set('started_at', backupStartedAt)
+              .value();
+            return BackupManager.registerBnRStatusPoller({
+              operation: CONST.OPERATION_TYPE.BACKUP,
+              type: backup.type,
+              trigger: backup.trigger
+            }, instanceInfo);
+          })
+          .then(agent_ip => {
+            registeredStatusPoller = true;
+            return this.agent.startBackup(agent_ip, backup, vms);
+          })
+          .then(() => {
+            backupStarted = true;
+            let put_ret = this.backupStore.putFile(data);
+            logger.info(put_ret);
+            return data;
+          });
+      })
+      .then(backup_options =>
+        eventmesh.apiServerClient.updateOperationResponse({
+          resourceId: opts.instance_guid,
+          operationName: CONST.APISERVER.ANNOTATION_NAMES.BACKUP,
+          operationType: CONST.APISERVER.ANNOTATION_TYPES.BACKUP,
+          operationId: result.backup_guid,
+          value: backup_options
+        })
+        .then(() =>
+          eventmesh.apiServerClient.updateOperationState({
```

Comment:
cant both operationResponse and state updated in 1 call? why do we need two calls here? May be poller might just update state of operation, but here response and state can be updated as one right?